File: apps/qt-ai-cli/src/commands/export.ts
import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { dirname, join, resolve } from "node:path";
import { naturalChapterCompare } from "@/lib/ai-story";
import { loadState, storyPaths } from "../story-fs.ts";

export interface ExportOptions {
  from?: string;
  to?: string;
  out?: string;
}

export interface ExportResult {
  outPath: string;
  ids: string[];
  /** Chương trong khoảng nhưng chưa done (queued/error/skipped) — file gộp bị hổng ở đó. */
  gaps: string[];
}

/**
 * Gộp out/<id>.txt của các chương done trong khoảng [from..to] thành một file,
 * mỗi chương cách nhau đúng một dòng trống — cùng format với "Tải chương" của web.
 * Mặc định khoảng = từ chương done nhỏ nhất tới lớn nhất; file ra export/<from>-<to>.txt.
 */
export function runExport(root: string, options: ExportOptions = {}): ExportResult {
  const paths = storyPaths(resolve(root));
  const state = loadState(paths);
  const all = Object.keys(state.chapters).sort(naturalChapterCompare);
  const done = all.filter((id) => state.chapters[id]!.status === "done");
  if (done.length === 0) throw new Error("Chưa có chương nào done để export.");

  const from = options.from ?? done[0]!;
  const to = options.to ?? done[done.length - 1]!;
  for (const bound of [from, to]) {
    if (!state.chapters[bound]) throw new Error(`Không có chương ${bound} trong state.json.`);
  }
  if (naturalChapterCompare(from, to) > 0) throw new Error(`Khoảng ngược: --from ${from} sau --to ${to}.`);

  const inRange = all.filter(
    (id) => naturalChapterCompare(id, from) >= 0 && naturalChapterCompare(id, to) <= 0,
  );
  const ids = inRange.filter((id) => state.chapters[id]!.status === "done");
  const gaps = inRange.filter((id) => state.chapters[id]!.status !== "done");
  if (ids.length === 0) throw new Error(`Không có chương done nào trong khoảng ${from}..${to}.`);

  const merged = ids
    .map((id) => {
      const file = join(paths.outDir, `${id}.txt`);
      if (!existsSync(file)) throw new Error(`Chương ${id} done nhưng thiếu ${file}.`);
      return readFileSync(file, "utf8").replace(/\s+$/u, "");
    })
    .filter(Boolean)
    .join("\n\n");

  const outPath = options.out
    ? resolve(options.out)
    : join(paths.root, "export", `${from}-${to}.txt`);
  mkdirSync(dirname(outPath), { recursive: true });
  writeFileSync(outPath, `${merged}\n`, "utf8");
  return { outPath, ids, gaps };
}

File: crates/qt-ai-core/src/commands/export.rs
use crate::error::{CoreError, Result};
use crate::story::natural_chapter_compare;
use crate::story_fs::{load_state, read_text, resolve_root, story_paths, write_text, ChapterStatus};
use std::cmp::Ordering;
use std::fs;
use std::path::{Path, PathBuf};

#[derive(Debug, Default, Clone)]
pub struct ExportOptions {
    pub from: Option<String>,
    pub to: Option<String>,
    pub out: Option<PathBuf>,
}

#[derive(Debug)]
pub struct ExportResult {
    pub out_path: PathBuf,
    pub ids: Vec<String>,
    /// Chương trong khoảng nhưng chưa done — file gộp bị hổng ở đó.
    pub gaps: Vec<String>,
}

/// Gộp out/<id>.txt của chương done trong [from..to], mỗi chương cách đúng một dòng trống (format
/// "Tải chương" của web). Mặc định khoảng = done đầu..done cuối; file ra export/<from>-<to>.txt.
pub fn run_export(root: &Path, options: &ExportOptions) -> Result<ExportResult> {
    let paths = story_paths(&resolve_root(root));
    let state = load_state(&paths)?;
    let mut all: Vec<String> = state.chapters.keys().cloned().collect();
    all.sort_by(|a, b| natural_chapter_compare(a, b));
    let done: Vec<&String> = all.iter().filter(|id| state.chapters[*id].status == ChapterStatus::Done).collect();
    let (Some(first_done), Some(last_done)) = (done.first(), done.last()) else {
        return Err(CoreError::InvalidState("Chưa có chương nào done để export.".to_string()));
    };
    let from = options.from.clone().unwrap_or_else(|| (*first_done).clone());
    let to = options.to.clone().unwrap_or_else(|| (*last_done).clone());
    for bound in [&from, &to] {
        if !state.chapters.contains_key(bound) {
            return Err(CoreError::StoryNotFound(format!("Không có chương {bound} trong state.json.")));
        }
    }
    if natural_chapter_compare(&from, &to) == Ordering::Greater {
        return Err(CoreError::InvalidState(format!("Khoảng ngược: --from {from} sau --to {to}.")));
    }
    let in_range: Vec<&String> = all
        .iter()
        .filter(|id| {
            natural_chapter_compare(id, &from) != Ordering::Less && natural_chapter_compare(id, &to) != Ordering::Greater
        })
        .collect();
    let ids: Vec<String> = in_range
        .iter()
        .filter(|id| state.chapters[**id].status == ChapterStatus::Done)
        .map(|id| (*id).clone())
        .collect();
    let gaps: Vec<String> = in_range
        .iter()
        .filter(|id| state.chapters[**id].status != ChapterStatus::Done)
        .map(|id| (*id).clone())
        .collect();
    if ids.is_empty() {
        return Err(CoreError::InvalidState(format!("Không có chương done nào trong khoảng {from}..{to}.")));
    }
    let mut parts = Vec::new();
    for id in &ids {
        let file = paths.out_dir.join(format!("{id}.txt"));
        if !file.exists() {
            return Err(CoreError::InvalidState(format!("Chương {id} done nhưng thiếu {}.", file.display())));
        }
        let text = read_text(&file)?;
        let trimmed = text.trim_end();
        if !trimmed.is_empty() {
            parts.push(trimmed.to_string());
        }
    }
    let out_path = match &options.out {
        Some(out) => resolve_root(out),
        None => paths.root.join("export").join(format!("{from}-{to}.txt")),
    };
    if let Some(parent) = out_path.parent() {
        fs::create_dir_all(parent).map_err(CoreError::io(parent))?;
    }
    write_text(&out_path, &format!("{}\n", parts.join("\n\n")))?;
    Ok(ExportResult { out_path, ids, gaps })
}
